File: tools/davinci/corpus-diff.mjs
#!/usr/bin/env node
// Corpus baseline diff gate (Davinci P0-5, suite TS-11).
//
// Runs a fresh whole-corpus sweep through the same reduction as
// `tools/davinci/corpus-baseline.mjs` and compares it against the committed
// baseline `tests/_fixtures/davinci-baseline.json`, reporting per-surface
// per-project drift.
//
// Exit 0 requires BOTH TS-11 conditions:
//   (a) zero drift between the fresh run and the committed baseline, and
//   (b) the scope proof holds — the baseline artifact exists and both the
//       baseline and the fresh run cover every project in
//       tests/_fixtures/vue-ecosystem-fixtures.json on every gated surface
//       with a nonzero total file count (a zero-file run fails).
// A missing baseline, a scope shortfall, or any drift exits 1 with the
// exact reasons.
//
// Usage:
//   node tools/davinci/corpus-diff.mjs [options]
//     --surface <s[,s]>  gate only these surfaces (compiler, formatter,
//                        linter, typechecker); the committed baseline is
//                        still validated against the full manifest scope
//     --shards <n>       parallel shard processes (default 4)
//     --vize-bin <path>  vize executable (default target/release/vize)
//     --baseline <path>  baseline artifact (default the committed path)
//     --write-fresh <p>  also write the fresh run's artifact to this path
//     --timeout-ms <n>   per-run timeout forwarded to the harness
//     --keep-raw         keep raw shard reports under .vize/davinci-corpus/
//
// Node builtins only.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";

import {
  buildArtifact,
  diffRows,
  renderArtifact,
  verifyScope,
} from "./lib/corpus-baseline-artifact.mjs";
import {
  BASELINE_PATH,
  SURFACES,
  UNSTABLE_REL,
  loadManifest,
  loadUnstableRows,
  parseSurfaceFilter,
} from "./lib/corpus-baseline-contract.mjs";
import {
  cleanupScratch,
  reduceShards,
  resolveVizeBin,
  runMatrix,
  scratchRoot,
} from "./lib/corpus-baseline-run.mjs";
import { assertFixturesPristine, cleanFixtures } from "./lib/corpus-fixture-hygiene.mjs";
import { repoRoot } from "./lib/paths.mjs";

function parseArgs(argv) {
  const args = {
    shards: 4,
    vizeBin: null,
    baseline: null,
    writeFresh: null,
    timeoutMs: null,
    keepRaw: false,
    surfaces: [],
    cleanFixtures: false,
    allowDirtyFixtures: false,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    const value = () => {
      if (argv[index + 1] == null) throw new Error(`${arg} requires a value`);
      return argv[++index];
    };
    if (arg === "--shards") args.shards = positiveInteger(value(), arg);
    else if (arg === "--vize-bin") args.vizeBin = value();
    else if (arg === "--baseline") args.baseline = value();
    else if (arg === "--write-fresh") args.writeFresh = value();
    else if (arg === "--timeout-ms") args.timeoutMs = positiveInteger(value(), arg);
    else if (arg === "--keep-raw") args.keepRaw = true;
    else if (arg === "--clean-fixtures") args.cleanFixtures = true;
    else if (arg === "--allow-dirty-fixtures") args.allowDirtyFixtures = true;
    else if (arg === "--surface") {
      args.surfaces.push(
        ...value()
          .split(",")
          .map((part) => part.trim())
          .filter(Boolean),
      );
    } else if (arg === "--help" || arg === "-h") {
      process.stdout.write(
        "usage: node tools/davinci/corpus-diff.mjs [--surface s[,s]] [--shards n] [--vize-bin path] [--baseline path] [--write-fresh path] [--timeout-ms n] [--keep-raw] [--clean-fixtures] [--allow-dirty-fixtures]\n",
      );
      process.exit(0);
    } else throw new Error(`unknown argument: ${arg}`);
  }
  return args;
}

function positiveInteger(raw, name) {
  const parsed = Number(raw);
  if (!Number.isInteger(parsed) || parsed <= 0) {
    throw new Error(`${name} must be a positive integer`);
  }
  return parsed;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const surfaces = parseSurfaceFilter(args.surfaces);
  const manifest = loadManifest();
  const baselinePath = args.baseline == null ? BASELINE_PATH : path.resolve(args.baseline);
  const baselineLabel = path.relative(repoRoot, baselinePath);
  const log = (line) => process.stdout.write(`${line}\n`);

  if (args.cleanFixtures) {
    const removed = cleanFixtures();
    log(`corpus-diff: cleaned ${removed} materialized node_modules from the fixtures`);
  }
  assertFixturesPristine(fail, { allowMaterialized: args.allowDirtyFixtures });

  if (!existsSync(baselinePath)) {
    fail([
      `baseline artifact is missing: ${baselineLabel}`,
      "generate it with: node tools/davinci/corpus-baseline.mjs",
    ]);
  }
  let baseline;
  try {
    baseline = JSON.parse(readFileSync(baselinePath, "utf8"));
  } catch (error) {
    fail([`baseline artifact is not valid JSON: ${baselineLabel} (${error.message})`]);
  }
  const baselineScopeFailures = verifyScope(baseline, manifest, SURFACES, "committed baseline");
  if (baselineScopeFailures.length > 0) {
    fail(["scope proof failed for the committed baseline:", ...baselineScopeFailures]);
  }
  const vizeBin = resolveVizeBin(args.vizeBin);

  log(
    `corpus-diff: ${manifest.projects.length} projects, surfaces [${surfaces.join(", ")}], ${args.shards} parallel shards`,
  );
  const startedAt = Date.now();
  const scratchDir = scratchRoot(`diff-${process.pid}`);
  let shardDirs;
  try {
    shardDirs = await runMatrix({
      shards: args.shards,
      vizeBin,
      tools: surfaces,
      scratchDir,
      timeoutMs: args.timeoutMs,
      log,
    });
  } catch (error) {
    log(`raw shard reports kept for debugging: ${path.relative(repoRoot, scratchDir)}`);
    throw error;
  }
  const freshRows = reduceShards(shardDirs, surfaces);
  const fresh = buildArtifact(freshRows, manifest);
  if (args.writeFresh != null) {
    const freshPath = path.resolve(args.writeFresh);
    mkdirSync(path.dirname(freshPath), { recursive: true });
    writeFileSync(freshPath, renderArtifact(fresh));
    log(`wrote fresh artifact: ${path.relative(repoRoot, freshPath)}`);
  }
  if (!args.keepRaw) cleanupScratch(scratchDir);
  else log(`raw shard reports kept: ${path.relative(repoRoot, scratchDir)}`);

  const freshScopeFailures = verifyScope(fresh, manifest, surfaces, "fresh run");
  const baselineRows = baseline.rows.filter((row) => surfaces.includes(row.surface));
  const allDrift = diffRows(baselineRows, freshRows);
  const unstableKeys = new Set(
    loadUnstableRows(manifest).map((row) => `${row.surface} ${row.project}`),
  );
  const drift = allDrift.filter(
    (record) =>
      record.kind !== "changed" || !unstableKeys.has(`${record.surface} ${record.project}`),
  );
  const unstableDrift = allDrift.filter((record) => !drift.includes(record));
  const elapsedSeconds = Math.round((Date.now() - startedAt) / 1000);

  for (const record of unstableDrift) {
    log(
      `unstable (filed in ${UNSTABLE_REL}, not gating): ${record.surface}/${record.project} ${record.baseline_hash.slice(0, 12)} -> ${record.fresh_hash.slice(0, 12)}`,
    );
  }
  if (drift.length > 0) {
    log(`drift: ${drift.length} row(s) differ from ${baselineLabel}`);
    for (const record of drift) {
      if (record.kind === "changed") {
        const fileNote =
          record.baseline_file_count === record.fresh_file_count
            ? `${record.fresh_file_count} files`
            : `files ${record.baseline_file_count} -> ${record.fresh_file_count}`;
        log(
          `  changed  ${record.surface}/${record.project} (${fileNote}) ${record.baseline_hash.slice(0, 12)} -> ${record.fresh_hash.slice(0, 12)}`,
        );
      } else {
        log(`  ${record.kind}  ${record.surface}/${record.project}`);
      }
    }
    const bySurface = new Map();
    for (const record of drift) {
      bySurface.set(record.surface, (bySurface.get(record.surface) ?? 0) + 1);
    }
    log(
      `drift by surface: ${[...bySurface.entries()]
        .map(([surface, count]) => `${surface}=${count}`)
        .join(", ")}`,
    );
  }
  if (freshScopeFailures.length > 0) {
    log("scope proof failed for the fresh run:");
    for (const reason of freshScopeFailures) log(`  ${reason}`);
  }
  if (drift.length > 0 || freshScopeFailures.length > 0) {
    log(`corpus-diff: FAIL in ${elapsedSeconds}s`);
    process.exit(1);
  }
  const filterNote =
    surfaces.length === SURFACES.length ? "" : ` [surface filter: ${surfaces.join(", ")}]`;
  const unstableNote =
    unstableDrift.length === 0
      ? ""
      : ` (${unstableDrift.length} filed unstable row(s) drifted without gating)`;
  log(
    `corpus-diff: PASS in ${elapsedSeconds}s — zero gating drift across ${freshRows.length} rows (${fresh.scope.projects_run} projects x ${fresh.scope.surfaces_per_project} surfaces, ${fresh.scope.total_file_count} files); scope proof matches ${manifest.projects.length}-project manifest${filterNote}${unstableNote}`,
  );
}

function fail(reasons) {
  for (const reason of reasons) process.stderr.write(`${reason}\n`);
  process.exit(1);
}

main().catch((error) => {
  process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
  process.exit(1);
});

File: tools/commands/davinci/corpus-diff.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [package]
//! edition = "2024"
//!
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! sha2 = "0.10"
//! ```

use serde_json::Value;
use std::{
    collections::BTreeSet,
    env,
    path::PathBuf,
    process::ExitCode,
    time::{Instant, SystemTime, UNIX_EPOCH},
};

#[path = "../../rust/common.rs"]
mod common;
#[path = "../../rust/davinci_corpus.rs"]
mod davinci_corpus;

#[derive(Debug)]
struct Args {
    shards: usize,
    vize_bin: Option<PathBuf>,
    baseline: Option<PathBuf>,
    write_fresh: Option<PathBuf>,
    timeout_ms: Option<u64>,
    keep_raw: bool,
    surfaces: Vec<String>,
    clean_fixtures: bool,
    allow_dirty_fixtures: bool,
}

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let root = common::repo_root()?;
    let args = parse_args(env::args().skip(1).collect())?;
    let surfaces = davinci_corpus::surface_filter(&args.surfaces)?;
    let manifest = davinci_corpus::load_manifest(&root)?;
    let baseline_path = absolutize(
        &root,
        args.baseline
            .unwrap_or_else(|| root.join(davinci_corpus::BASELINE_REL)),
    );
    let baseline_label = common::relative_path(&root, &baseline_path);

    if args.clean_fixtures {
        let removed = davinci_corpus::clean_fixtures(&root)?;
        println!("corpus-diff: cleaned {removed} materialized node_modules from the fixtures");
    }
    davinci_corpus::assert_fixtures_pristine(&root, args.allow_dirty_fixtures)?;

    if !baseline_path.exists() {
        return Err(format!(
            "baseline artifact is missing: {baseline_label}\ngenerate it with: rust-script tools/commands/davinci/corpus-baseline.rs"
        ));
    }
    let baseline = common::read_json(&baseline_path)?;
    let all_surfaces = davinci_corpus::surface_filter(&Vec::new())?;
    let baseline_scope_failures =
        davinci_corpus::verify_scope(&baseline, &manifest, &all_surfaces, "committed baseline")?;
    if !baseline_scope_failures.is_empty() {
        return Err(format!(
            "scope proof failed for the committed baseline:\n{}",
            baseline_scope_failures.join("\n")
        ));
    }
    let vize_bin = davinci_corpus::resolve_vize_bin(&root, args.vize_bin)?;

    println!(
        "corpus-diff: {} projects, surfaces [{}], {} parallel shards",
        manifest["projects"].as_array().map_or(0, Vec::len),
        surfaces.join(", "),
        args.shards
    );
    let started = Instant::now();
    let scratch_dir = davinci_corpus::scratch_root(&root, &format!("diff-{}", process_slug()));
    let shard_dirs = match davinci_corpus::run_matrix(
        &root,
        args.shards,
        &vize_bin,
        &surfaces,
        &scratch_dir,
        args.timeout_ms,
    ) {
        Ok(shard_dirs) => shard_dirs,
        Err(error) => {
            println!(
                "raw shard reports kept for debugging: {}",
                common::relative_path(&root, &scratch_dir)
            );
            return Err(error);
        }
    };
    let fresh_rows = davinci_corpus::reduce_shards(&root, &shard_dirs, &surfaces)?;
    let fresh = davinci_corpus::build_artifact(&fresh_rows, &manifest)?;
    if let Some(write_fresh) = args.write_fresh {
        let fresh_path = absolutize(&root, write_fresh);
        common::write_json_pretty(&fresh_path, &fresh)?;
        println!(
            "wrote fresh artifact: {}",
            common::relative_path(&root, &fresh_path)
        );
    }
    if !args.keep_raw {
        davinci_corpus::cleanup_scratch(&scratch_dir)?;
    } else {
        println!(
            "raw shard reports kept: {}",
            common::relative_path(&root, &scratch_dir)
        );
    }

    let fresh_scope_failures =
        davinci_corpus::verify_scope(&fresh, &manifest, &surfaces, "fresh run")?;
    let baseline_rows = baseline
        .get("rows")
        .and_then(Value::as_array)
        .ok_or_else(|| "baseline rows must be an array".to_string())?
        .iter()
        .filter(|row| {
            row.get("surface")
                .and_then(Value::as_str)
                .is_some_and(|surface| surfaces.iter().any(|item| item == surface))
        })
        .cloned()
        .collect::<Vec<_>>();
    let all_drift = davinci_corpus::diff_rows(&baseline_rows, &fresh_rows);
    let unstable_keys = davinci_corpus::load_unstable_rows(&root, &manifest)?
        .iter()
        .filter_map(|row| {
            Some(format!(
                "{} {}",
                row.get("surface")?.as_str()?,
                row.get("project")?.as_str()?
            ))
        })
        .collect::<BTreeSet<_>>();
    let mut drift = Vec::new();
    let mut unstable_drift = Vec::new();
    for record in all_drift {
        let key = format!(
            "{} {}",
            record.get("surface").and_then(Value::as_str).unwrap_or(""),
            record.get("project").and_then(Value::as_str).unwrap_or("")
        );
        if record.get("kind").and_then(Value::as_str) == Some("changed")
            && unstable_keys.contains(&key)
        {
            unstable_drift.push(record);
        } else {
            drift.push(record);
        }
    }
    let elapsed_seconds = started.elapsed().as_secs();

    for record in &unstable_drift {
        println!(
            "unstable (filed in {}, not gating): {}/{} {} -> {}",
            davinci_corpus::UNSTABLE_REL,
            record.get("surface").and_then(Value::as_str).unwrap_or("?"),
            record.get("project").and_then(Value::as_str).unwrap_or("?"),
            prefix12(
                record
                    .get("baseline_hash")
                    .and_then(Value::as_str)
                    .unwrap_or("")
            ),
            prefix12(
                record
                    .get("fresh_hash")
                    .and_then(Value::as_str)
                    .unwrap_or("")
            )
        );
    }
    if !drift.is_empty() {
        println!("drift: {} row(s) differ from {baseline_label}", drift.len());
        for record in &drift {
            let kind = record.get("kind").and_then(Value::as_str).unwrap_or("?");
            let surface = record.get("surface").and_then(Value::as_str).unwrap_or("?");
            let project = record.get("project").and_then(Value::as_str).unwrap_or("?");
            if kind == "changed" {
                let baseline_file_count = record
                    .get("baseline_file_count")
                    .and_then(Value::as_u64)
                    .unwrap_or(0);
                let fresh_file_count = record
                    .get("fresh_file_count")
                    .and_then(Value::as_u64)
                    .unwrap_or(0);
                let file_note = if baseline_file_count == fresh_file_count {
                    format!("{fresh_file_count} files")
                } else {
                    format!("files {baseline_file_count} -> {fresh_file_count}")
                };
                println!(
                    "  changed  {surface}/{project} ({file_note}) {} -> {}",
                    prefix12(
                        record
                            .get("baseline_hash")
                            .and_then(Value::as_str)
                            .unwrap_or("")
                    ),
                    prefix12(
                        record
                            .get("fresh_hash")
                            .and_then(Value::as_str)
                            .unwrap_or("")
                    )
                );
            } else {
                println!("  {kind}  {surface}/{project}");
            }
        }
        let mut by_surface = std::collections::BTreeMap::<String, usize>::new();
        for record in &drift {
            let surface = record
                .get("surface")
                .and_then(Value::as_str)
                .unwrap_or("?")
                .to_string();
            *by_surface.entry(surface).or_default() += 1;
        }
        println!(
            "drift by surface: {}",
            by_surface
                .into_iter()
                .map(|(surface, count)| format!("{surface}={count}"))
                .collect::<Vec<_>>()
                .join(", ")
        );
    }
    if !fresh_scope_failures.is_empty() {
        println!("scope proof failed for the fresh run:");
        for reason in &fresh_scope_failures {
            println!("  {reason}");
        }
    }
    if !drift.is_empty() || !fresh_scope_failures.is_empty() {
        println!("corpus-diff: FAIL in {elapsed_seconds}s");
        return Err("corpus-diff failed".to_string());
    }
    let expected_fresh_comparisons =
        davinci_corpus::expected_comparison_count(&manifest, &surfaces)?;
    let filter_note = if surfaces.len() == all_surfaces.len() {
        String::new()
    } else {
        format!(" [surface filter: {}]", surfaces.join(", "))
    };
    let unstable_note = if unstable_drift.is_empty() {
        String::new()
    } else {
        format!(
            " ({} filed unstable row(s) drifted without gating)",
            unstable_drift.len()
        )
    };
    println!(
        "corpus-diff: PASS in {elapsed_seconds}s - zero gating drift across {}/{} comparisons ({} projects x {} surfaces, {} files); scope proof matches {}-project manifest{filter_note}{unstable_note}",
        fresh_rows.len(),
        expected_fresh_comparisons,
        fresh["scope"]["projects_run"],
        fresh["scope"]["surfaces_per_project"],
        fresh["scope"]["total_file_count"],
        manifest["projects"].as_array().map_or(0, Vec::len)
    );
    Ok(())
}

fn parse_args(argv: Vec<String>) -> Result<Args, String> {
    let mut args = Args {
        shards: 4,
        vize_bin: None,
        baseline: None,
        write_fresh: None,
        timeout_ms: None,
        keep_raw: false,
        surfaces: Vec::new(),
        clean_fixtures: false,
        allow_dirty_fixtures: false,
    };
    let mut index = 0;
    while index < argv.len() {
        let arg = &argv[index];
        let value = |index: &mut usize| -> Result<String, String> {
            *index += 1;
            argv.get(*index)
                .cloned()
                .ok_or_else(|| format!("{arg} requires a value"))
        };
        match arg.as_str() {
            "--shards" => {
                args.shards = davinci_corpus::positive_integer(&value(&mut index)?, arg)? as usize
            }
            "--vize-bin" => args.vize_bin = Some(PathBuf::from(value(&mut index)?)),
            "--baseline" => args.baseline = Some(PathBuf::from(value(&mut index)?)),
            "--write-fresh" => args.write_fresh = Some(PathBuf::from(value(&mut index)?)),
            "--timeout-ms" => {
                args.timeout_ms = Some(davinci_corpus::positive_integer(&value(&mut index)?, arg)?)
            }
            "--keep-raw" => args.keep_raw = true,
            "--clean-fixtures" => args.clean_fixtures = true,
            "--allow-dirty-fixtures" => args.allow_dirty_fixtures = true,
            "--surface" => args
                .surfaces
                .extend(davinci_corpus::split_csv(&value(&mut index)?)),
            "--help" | "-h" => {
                print_help();
                std::process::exit(0);
            }
            _ => return Err(format!("unknown argument: {arg}")),
        }
        index += 1;
    }
    Ok(args)
}

fn print_help() {
    println!(
        "usage: rust-script tools/commands/davinci/corpus-diff.rs [--surface s[,s]] [--shards n] [--vize-bin path] [--baseline path] [--write-fresh path] [--timeout-ms n] [--keep-raw] [--clean-fixtures] [--allow-dirty-fixtures]"
    );
}

fn absolutize(root: &std::path::Path, path: PathBuf) -> PathBuf {
    if path.is_absolute() {
        path
    } else {
        root.join(path)
    }
}

fn prefix12(value: &str) -> &str {
    value.get(..12).unwrap_or(value)
}

fn process_slug() -> String {
    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_millis())
        .unwrap_or(0);
    format!("{}-{millis}", std::process::id())
}
